tests/request_processing/test_grundsteuer_request_controller.py

```diff
@@ -0,0 +1,57 @@
+import base64
+from unittest.mock import patch, MagicMock
+from xml.etree import ElementTree
+
+import pytest
+
+from erica.pyeric.pyeric_response import PyericResponse
+from erica.request_processing.grundsteuer_request_controller import GrundsteuerRequestController
+from tests.sample_data import create_grundsteuer
+from tests.utils import missing_cert, missing_pyeric_lib
+
+
+@pytest.fixture
+def valid_grundsteuer_request_controller():
+    grundsteuer_input = create_grundsteuer()
+    return GrundsteuerRequestController(grundsteuer_input)
+
+
+class TestIsTestmerkerUsed:
+    def test_returns_true(self, valid_grundsteuer_request_controller):
+        result = valid_grundsteuer_request_controller._is_testmerker_used()
+        assert result is True
+
+
+@pytest.mark.skipif(missing_cert(), reason="skipped because of missing cert.pfx; see pyeric/README.md")
+@pytest.mark.skipif(missing_pyeric_lib(), reason="skipped because of missing eric lib; see pyeric/README.md")
+class TestGenerateFullXml:
+    def test_returns_valid_xml(self, valid_grundsteuer_request_controller):
+        resulting_xml = valid_grundsteuer_request_controller.generate_full_xml(use_testmerker=True)
+        try:
+            ElementTree.fromstring(resulting_xml)
+        except ElementTree.ParseError as e:
+            return pytest.fail("Did not result in a valid xml: \n" + e.msg)
+
+    def test_returned_xml_includes_transfer_header(self, valid_grundsteuer_request_controller):
+        resulting_xml = valid_grundsteuer_request_controller.generate_full_xml(use_testmerker=True)
+        assert "<TransferHeader" in resulting_xml
+
+    def test_returned_xml_includes_nutzdaten(self, valid_grundsteuer_request_controller):
+        resulting_xml = valid_grundsteuer_request_controller.generate_full_xml(use_testmerker=True)
+        assert "<Nutzdaten" in resulting_xml
+
+    def test_returned_xml_includes_nutzdaten_header(self, valid_grundsteuer_request_controller):
+        resulting_xml = valid_grundsteuer_request_controller.generate_full_xml(use_testmerker=True)
+        assert "<NutzdatenHeader" in resulting_xml
```

Comment:
I don't think I wanna do that. I'm less and less a fan of making sure a function is called and testing for the data..when does it stop? I think this could be solved by the test that Shahrooz suggested?

> Apart from that, there's one extra test case that I'd love us to have: complete json as input, complete xml as expected output, each residing in its own separate file, then performing the whole conversion and asserting the equality of the generated output with the expected one (I'm sure there are ways/tools to compare the xmls without the need to perform a manual text comparison). It would provide a great (and easy to grasp) way to verify that everything is working correctly.

Reply:
> I'm less and less a fan of making sure a function is called and testing for the data..when does it stop?

Imo it is only important to make sure that we test all potential input data classes at some point. When we just pass the data through to underlying functions and we do blackbox testing then we need to test all the different inputs/output pairs again on all levels while we do not have to do this when we use white ox testing when a function only calls other functions.

Personally, I like blackbox testing more because we do not have to change tests that much, if we test all the input pairs.

Anyways, I like the test Sharooz mentioned.